### ISSUE TITLE
Port to Marshmallow 3

### ISSUE THREAD

**Issue by lafrech** [Scille/umongo#129]
Creating this issue to discuss things related to Marshmallow 3.

Marshmallow 3 will introduce breaking changes. Assuming we want to keep compatibility with Marshmallow 2, we'd need to introduce conditional code to adapt to both versions and test on both v2 and v3. This can be started before Marshmallow 3.0.0 is out.

Regarding the tests, here's how I did it on another lib: https://github.com/Nobatek/flask-rest-api/commit/cd23fa2ccfba47b51ec776ff47c74b447e02b378. Inspired from `mongoengine` testing on several `pymongo` versions. Obviously this means the test will break each time the dev branch introduces breaking changes.

Another approach would be to work on a separate branch and only merge when Marshmallow 3 is stable.

In the meantime, we can also help closing v3 issues in Marshmallow to speed up the process: https://github.com/marshmallow-code/marshmallow/issues?q=is%3Aopen+is%3Aissue+milestone%3A3.0.

**Comment by touilleMan** [comment]
What kind of changes introduce marshmallow 3 ?

> Assuming we want to keep compatibility with Marshmallow 2

Not sure if it's something wanted as compatibility means more work...

It obviously depend on how much work is involved, but I say we could release a new major version with marshmallow 3 and from there backport really important patches to the actual major version (which would become the marshmallow 2 support branch)

**Comment by lafrech** [comment]
> What kind of changes introduce marshmallow 3 ?

See [changelog](https://github.com/marshmallow-code/marshmallow/blob/dev/CHANGELOG.rst) and [upgrading to 3.0](https://github.com/marshmallow-code/marshmallow/blob/dev/docs/upgrading.rst) for what is done already. And [issues milestones 3.0](https://github.com/marshmallow-code/marshmallow/issues?q=is%3Aopen+is%3Aissue+milestone%3A3.0) for what is to come.

All schemas are now strict, and only data is returned. Errors trigger an exception containing error messages. This and the other changes should be easy fixes in umongo. I even hope this can be done with a single umongo version supporting both marshmallow versions, with a if `MARSHMALLOW_2` test in a few places. This would be less work than backporting patches IMHO.

Going marshmallow 3 and dropping marshmallow 2 support would be an issue for people using other marshmallow based libraries that wouldn't be ported yet. I'm thinking of webargs and apispec, but there may be more. Once those two are ported, I don't mind about marshmallow 2 anymore. Considering the changes we're talking about, apispec could be a painful port, but we'll see.

Let's be optimistic and aim at single umongo version supporting MA2 and MA3. We can rediscuss this if problems arise.

**Comment by touilleMan** [comment]
> Let's be optimistic and aim at single umongo version supporting MA2 and MA3. We can rediscuss this if problems arise.

I like this 😄 

One thing though, I think we should wait for marshmallow 3.0 stable release before any work.
Reason is I've started umongo right when marshmallow 2.0 was around the corner, but got plenty of annoying breaks due to their API being changed constantly (what did I expect ? that's what beta are for anyway !)

**Comment by lafrech** [comment]
OK.

I've been following the recent moves on Marshmallow 3, and I've got an eye on the pipe (issues / PRs marked v3). From what I've seen, there's not much to be worried about for umongo. I might begin work before the release, but this is not a priority for me right now anyway.

**Comment by lafrech** [comment]
I've been working on this (https://github.com/Scille/umongo/pull/154). There are not so many changes in umongo code. More in the tests.

I'm reconsidering supporting both versions. It would be doable, but considering the number of changes in umongo last year, it seems simpler to issue umongo v2 with marshmallow 3 support (and dropping Python 3.4) and keep a umongo 1.x maintenance branch. There shouldn't be so many fixes to backport.

In any case, let's wait until marshmallow 3.0.0 is out. I don't think we'll have it for Christmas.

**Comment by gerasim13** [comment]
Maybe it will be useful for you to look at https://github.com/lyft/toasted-marshmallow

**Comment by Cayke** [comment]
Any news on this?
